File: Alkitab/src/main/java/yuku/alkitab/base/ac/GotoActivity.java
package yuku.alkitab.base.ac;

import android.content.Intent;
import android.os.Bundle;
import android.support.design.widget.TabLayout;
import android.support.v4.app.Fragment;
import android.support.v4.app.FragmentManager;
import android.support.v4.app.FragmentPagerAdapter;
import android.support.v4.view.ViewPager;
import android.support.v7.app.ActionBar;
import android.support.v7.widget.Toolbar;
import android.view.Menu;
import android.view.MenuItem;
import android.view.View;
import android.view.inputmethod.InputMethodManager;
import yuku.afw.App;
import yuku.afw.V;
import yuku.afw.storage.Preferences;
import yuku.alkitab.base.ac.base.BaseActivity;
import yuku.alkitab.base.fr.GotoDialerFragment;
import yuku.alkitab.base.fr.GotoDirectFragment;
import yuku.alkitab.base.fr.GotoGridFragment;
import yuku.alkitab.base.fr.base.BaseGotoFragment;
import yuku.alkitab.base.storage.Prefkey;
import yuku.alkitab.debug.R;

public class GotoActivity extends BaseActivity implements BaseGotoFragment.GotoFinishListener {
	public static final String TAG = GotoActivity.class.getSimpleName();

	private static final String EXTRA_bookId = "bookId";
	private static final String EXTRA_chapter = "chapter";
	private static final String EXTRA_verse = "verse";

	private static final String INSTANCE_STATE_tab = "tab";

	public static class Result {
		public int bookId;
		public int chapter_1;
		public int verse_1;
	}

	public static Intent createIntent(int bookId, int chapter_1, int verse_1) {
		Intent res = new Intent(App.context, GotoActivity.class);
		res.putExtra(EXTRA_bookId, bookId);
		res.putExtra(EXTRA_chapter, chapter_1);
		res.putExtra(EXTRA_verse, verse_1);
		return res;
	}

	public static Result obtainResult(Intent data) {
		Result res = new Result();
		res.bookId = data.getIntExtra(EXTRA_bookId, -1);
		res.chapter_1 = data.getIntExtra(EXTRA_chapter, 0);
		res.verse_1 = data.getIntExtra(EXTRA_verse, 0);
		return res;
	}

	Toolbar toolbar;
	ViewPager viewPager;
	TabLayout tablayout;
	GotoPagerAdapter pagerAdapter;

	boolean okToHideKeyboard = false;

	int bookId;
	int chapter_1;
	int verse_1;

	@Override public void onCreate(Bundle savedInstanceState) {
		super.onCreate(savedInstanceState);
		setContentView(R.layout.activity_goto);

		bookId = getIntent().getIntExtra(EXTRA_bookId, -1);
		chapter_1 = getIntent().getIntExtra(EXTRA_chapter, 0);
		verse_1 = getIntent().getIntExtra(EXTRA_verse, 0);

		toolbar = V.get(this, R.id.toolbar);
		setSupportActionBar(toolbar);
		final ActionBar actionBar = getSupportActionBar();
		if (actionBar != null) {
			actionBar.setDisplayShowTitleEnabled(false);
			actionBar.setDisplayHomeAsUpEnabled(true);
			toolbar.setNavigationOnClickListener(v -> navigateUp());
		}

		// ViewPager and its adapters use support library fragments, so use getSupportFragmentManager.
		viewPager = V.get(this, R.id.viewPager);
		viewPager.setAdapter(pagerAdapter = new GotoPagerAdapter(getSupportFragmentManager()));
		viewPager.addOnPageChangeListener(new ViewPager.SimpleOnPageChangeListener() {
			@Override
			public void onPageSelected(int position) {
				if (okToHideKeyboard && position != 1) {
					final View editText = V.get(GotoActivity.this, R.id.tDirectReference);
					if (editText != null) {
						final InputMethodManager imm = (InputMethodManager) getSystemService(INPUT_METHOD_SERVICE);
						imm.hideSoftInputFromWindow(editText.getWindowToken(), InputMethodManager.HIDE_IMPLICIT_ONLY);
						imm.hideSoftInputFromWindow(editText.getWindowToken(), InputMethodManager.HIDE_NOT_ALWAYS);
					}
				}
			}
		});

		tablayout = V.get(this, R.id.tablayout);
		tablayout.setTabMode(TabLayout.MODE_SCROLLABLE);
		tablayout.setupWithViewPager(viewPager);

		if (savedInstanceState == null) {
			// get from preferences
			int tabUsed = Preferences.getInt(Prefkey.goto_last_tab, 0);
			if (tabUsed >= 1 && tabUsed <= 3) {
				viewPager.setCurrentItem(tabUsed - 1 /* to make it 0-based */, false);
			}

			if (tabUsed == 2) {
				viewPager.postDelayed(() -> {
					final View editText = V.get(GotoActivity.this, R.id.tDirectReference);
					if (editText != null) {
						InputMethodManager imm = (InputMethodManager) getSystemService(INPUT_METHOD_SERVICE);
						imm.showSoftInput(editText, InputMethodManager.SHOW_IMPLICIT);
					}
					okToHideKeyboard = true;
				}, 100);
			} else {
				okToHideKeyboard = true;
			}
		} else {
			viewPager.setCurrentItem(savedInstanceState.getInt(INSTANCE_STATE_tab, 0), false);
		}
	}

	@Override
	public boolean onCreateOptionsMenu(final Menu menu) {
		getMenuInflater().inflate(R.menu.activity_goto, menu);
		return true;
	}

	@Override
	public boolean onPrepareOptionsMenu(final Menu menu) {
		final MenuItem menuAskForVerse = menu.findItem(R.id.menuAskForVerse);
		final boolean val = Preferences.getBoolean(Prefkey.gotoAskForVerse, Prefkey.GOTO_ASK_FOR_VERSE_DEFAULT);
		menuAskForVerse.setChecked(val);

		return true;
	}

	@Override
	public boolean onOptionsItemSelected(final MenuItem item) {
		if (item.getItemId() == R.id.menuAskForVerse) {
			final boolean val = Preferences.getBoolean(Prefkey.gotoAskForVerse, Prefkey.GOTO_ASK_FOR_VERSE_DEFAULT);
			Preferences.setBoolean(Prefkey.gotoAskForVerse, !val);
			supportInvalidateOptionsMenu();
			return true;
		}

		return super.onOptionsItemSelected(item);
	}

	@Override protected void onSaveInstanceState(Bundle outState) {
		super.onSaveInstanceState(outState);
		outState.putInt(INSTANCE_STATE_tab, viewPager.getCurrentItem());
	}

	public class GotoPagerAdapter extends FragmentPagerAdapter {
		final int[] pageTitleResIds = {R.string.goto_tab_dialer_label, R.string.goto_tab_direct_label, R.string.goto_tab_grid_label};

		public GotoPagerAdapter(FragmentManager fm) {
			super(fm);
		}

		@Override
		public Fragment getItem(final int position) {
			final Fragment res;
			if (position == 0) {
				res = Fragment.instantiate(GotoActivity.this, GotoDialerFragment.class.getName(), GotoDialerFragment.createArgs(bookId, chapter_1, verse_1));
			} else if (position == 1) {
				res = Fragment.instantiate(GotoActivity.this, GotoDirectFragment.class.getName(), GotoDirectFragment.createArgs(bookId, chapter_1, verse_1));
			} else {
				res = Fragment.instantiate(GotoActivity.this, GotoGridFragment.class.getName(), GotoGridFragment.createArgs(bookId, chapter_1, verse_1));
			}
			return res;
		}

		@Override
		public int getCount() {
			return pageTitleResIds.length;
		}

		@Override
		public CharSequence getPageTitle(int position) {
			return getString(pageTitleResIds[position]);
		}
	}

	@Override public void onGotoFinished(int gotoTabUsed, int bookId, int chapter_1, int verse_1) {
		// store goto tab used for next time
		Preferences.setInt(Prefkey.goto_last_tab, gotoTabUsed);
		
		Intent data = new Intent();
		data.putExtra(EXTRA_bookId, bookId);
		data.putExtra(EXTRA_chapter, chapter_1);
		data.putExtra(EXTRA_verse, verse_1);
		setResult(RESULT_OK, data);
		finish();
	}
}

File: Alkitab/src/main/java/yuku/alkitab/base/ac/GotoActivity.kt
package yuku.alkitab.base.ac

import android.content.Intent
import android.os.Bundle
import android.view.Menu
import android.view.MenuItem
import android.view.View
import android.view.inputmethod.InputMethodManager
import androidx.activity.compose.setContent
import androidx.appcompat.widget.Toolbar
import androidx.fragment.app.Fragment
import androidx.fragment.app.FragmentManager
import androidx.fragment.app.FragmentPagerAdapter
import androidx.viewpager.widget.ViewPager
import com.google.android.material.tabs.TabLayout
import yuku.afw.App
import yuku.afw.storage.Preferences
import yuku.alkitab.base.ac.base.BaseActivity
import yuku.alkitab.base.compose.BibleAppTheme
import yuku.alkitab.base.compose.goto.GotoScreen
import yuku.alkitab.base.fr.GotoDialerFragment
import yuku.alkitab.base.fr.GotoDirectFragment
import yuku.alkitab.base.fr.GotoGridFragment
import yuku.alkitab.base.fr.base.BaseGotoFragment
import yuku.alkitab.base.settings.ExperimentalFlags
import yuku.alkitab.base.storage.GOTO_ASK_FOR_VERSE_DEFAULT
import yuku.alkitab.base.storage.Prefkey
import yuku.alkitab.debug.R

class GotoActivity : BaseActivity(), BaseGotoFragment.GotoFinishListener {

    class Result {
        @JvmField var bookId: Int = -1
        @JvmField var chapter_1: Int = 0
        @JvmField var verse_1: Int = 0
    }

    companion object {
        private const val EXTRA_bookId = "bookId"
        private const val EXTRA_chapter = "chapter"
        private const val EXTRA_verse = "verse"

        private const val INSTANCE_STATE_tab = "tab"

        @JvmStatic
        fun createIntent(bookId: Int, chapter_1: Int, verse_1: Int): Intent {
            return Intent(App.context, GotoActivity::class.java).apply {
                putExtra(EXTRA_bookId, bookId)
                putExtra(EXTRA_chapter, chapter_1)
                putExtra(EXTRA_verse, verse_1)
            }
        }

        @JvmStatic
        fun obtainResult(data: Intent): Result {
            return Result().apply {
                bookId = data.getIntExtra(EXTRA_bookId, -1)
                chapter_1 = data.getIntExtra(EXTRA_chapter, 0)
                verse_1 = data.getIntExtra(EXTRA_verse, 0)
            }
        }
    }

    private var useCompose = false

    // Legacy mode state. Only valid when [useCompose] is false.
    private var bookId: Int = -1
    private var chapter_1: Int = 0
    private var verse_1: Int = 0
    private var viewPager: ViewPager? = null
    private var okToHideKeyboard = false

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)

        bookId = intent.getIntExtra(EXTRA_bookId, -1)
        chapter_1 = intent.getIntExtra(EXTRA_chapter, 0)
        verse_1 = intent.getIntExtra(EXTRA_verse, 0)

        useCompose = ExperimentalFlags.useComposeGoto()

        if (useCompose) {
            setContent {
                BibleAppTheme {
                    GotoScreen(
                        initialBookId = bookId,
                        initialChapter_1 = chapter_1,
                        initialVerse_1 = verse_1,
                        onUp = { finish() },
                        onGotoFinished = { tab, bookId, chapter, verse ->
                            Preferences.setInt(Prefkey.goto_last_tab, tab.ordinal1Based)
                            val data = Intent().apply {
                                putExtra(EXTRA_bookId, bookId)
                                putExtra(EXTRA_chapter, chapter)
                                putExtra(EXTRA_verse, verse)
                            }
                            setResult(RESULT_OK, data)
                            finish()
                        },
                    )
                }
            }
        } else {
            setupLegacyUi(savedInstanceState)
        }
    }

    private fun setupLegacyUi(savedInstanceState: Bundle?) {
        setContentView(R.layout.activity_goto)

        val toolbar = findViewById<Toolbar>(R.id.toolbar)
        setSupportActionBar(toolbar)
        setupEdgeToEdgeDisplay(toolbar)
        supportActionBar?.apply {
            setDisplayShowTitleEnabled(false)
            setDisplayHomeAsUpEnabled(true)
        }
        toolbar.setNavigationOnClickListener { navigateUp() }

        val viewPager = findViewById<ViewPager>(R.id.viewPager)
        this.viewPager = viewPager
        viewPager.adapter = GotoPagerAdapter(supportFragmentManager)
        viewPager.addOnPageChangeListener(object : ViewPager.SimpleOnPageChangeListener() {
            override fun onPageSelected(position: Int) {
                if (okToHideKeyboard && position != 1) {
                    val editText = findViewById<View?>(R.id.tDirectReference)
                    if (editText != null) {
                        val imm = getSystemService(INPUT_METHOD_SERVICE) as InputMethodManager
                        imm.hideSoftInputFromWindow(editText.windowToken, InputMethodManager.HIDE_IMPLICIT_ONLY)
                        imm.hideSoftInputFromWindow(editText.windowToken, InputMethodManager.HIDE_NOT_ALWAYS)
                    }
                }
            }
        })

        val tablayout = findViewById<TabLayout>(R.id.tablayout)
        tablayout.tabMode = TabLayout.MODE_SCROLLABLE
        tablayout.setupWithViewPager(viewPager)

        if (savedInstanceState == null) {
            val tabUsed = Preferences.getInt(Prefkey.goto_last_tab, 0)
            if (tabUsed in 1..3) {
                viewPager.setCurrentItem(tabUsed - 1, false)
            }

            if (tabUsed == 2) {
                viewPager.postDelayed({
                    val editText = findViewById<View?>(R.id.tDirectReference)
                    if (editText != null) {
                        val imm = getSystemService(INPUT_METHOD_SERVICE) as InputMethodManager
                        imm.showSoftInput(editText, InputMethodManager.SHOW_IMPLICIT)
                    }
                    okToHideKeyboard = true
                }, 100)
            } else {
                okToHideKeyboard = true
            }
        } else {
            viewPager.setCurrentItem(savedInstanceState.getInt(INSTANCE_STATE_tab, 0), false)
        }
    }

    override fun onCreateOptionsMenu(menu: Menu): Boolean {
        if (useCompose) return super.onCreateOptionsMenu(menu)

        menuInflater.inflate(R.menu.activity_goto, menu)
        return true
    }

    override fun onPrepareOptionsMenu(menu: Menu): Boolean {
        if (!useCompose) {
            val menuAskForVerse = menu.findItem(R.id.menuAskForVerse)
            val v = Preferences.getBoolean(Prefkey.gotoAskForVerse, GOTO_ASK_FOR_VERSE_DEFAULT)
            menuAskForVerse.isChecked = v
        }
        return super.onPrepareOptionsMenu(menu)
    }

    override fun onOptionsItemSelected(item: MenuItem): Boolean {
        if (!useCompose && item.itemId == R.id.menuAskForVerse) {
            val v = Preferences.getBoolean(Prefkey.gotoAskForVerse, GOTO_ASK_FOR_VERSE_DEFAULT)
            Preferences.setBoolean(Prefkey.gotoAskForVerse, !v)
            invalidateOptionsMenu()
            return true
        }
        return super.onOptionsItemSelected(item)
    }

    override fun onSaveInstanceState(outState: Bundle) {
        super.onSaveInstanceState(outState)
        if (!useCompose) {
            viewPager?.let { outState.putInt(INSTANCE_STATE_tab, it.currentItem) }
        }
    }

    private inner class GotoPagerAdapter(fm: FragmentManager) : FragmentPagerAdapter(fm) {
        private val pageTitleResIds = intArrayOf(
            R.string.goto_tab_dialer_label,
            R.string.goto_tab_direct_label,
            R.string.goto_tab_grid_label,
        )

        override fun getItem(position: Int): Fragment {
            return when (position) {
                0 -> Fragment.instantiate(this@GotoActivity, GotoDialerFragment::class.java.name, GotoDialerFragment.createArgs(bookId, chapter_1, verse_1))
                1 -> Fragment.instantiate(this@GotoActivity, GotoDirectFragment::class.java.name, GotoDirectFragment.createArgs(bookId, chapter_1, verse_1))
                else -> Fragment.instantiate(this@GotoActivity, GotoGridFragment::class.java.name, GotoGridFragment.createArgs(bookId, chapter_1, verse_1))
            }
        }

        override fun getCount(): Int = pageTitleResIds.size

        override fun getPageTitle(position: Int): CharSequence = getString(pageTitleResIds[position])
    }

    override fun onGotoFinished(gotoTabUsed: Int, bookId: Int, chapter_1: Int, verse_1: Int) {
        Preferences.setInt(Prefkey.goto_last_tab, gotoTabUsed)

        val data = Intent().apply {
            putExtra(EXTRA_bookId, bookId)
            putExtra(EXTRA_chapter, chapter_1)
            putExtra(EXTRA_verse, verse_1)
        }
        setResult(RESULT_OK, data)
        finish()
    }
}
